File: typescript/src/security/pairing.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs"
import path from "node:path"
import type { BancoKeys, Identity } from "./keys.js"

export type Integracao = {
  bancoId: string
  keyid: string
  atlasPublicKey: string
  ambiente: string
}

export type PairInput = {
  atlasUrl: string
  token: string
  publicBaseUrl: string
  publicKey: string
  keyid: string
}

export async function pair(input: PairInput): Promise<Integracao> {
  const url = `${input.atlasUrl.replace(/\/$/, "")}/v1/integracao/parear`
  const response = await fetch(url, {
    method: "POST",
    headers: {
      authorization: `Bearer ${input.token}`,
      "content-type": "application/json",
    },
    body: JSON.stringify({
      public_key: input.publicKey,
      base_url: input.publicBaseUrl,
      keyid: input.keyid,
    }),
  })
  if (!response.ok) {
    const corpo = await response.text().catch(() => "")
    throw new Error(`pareamento falhou: HTTP ${response.status} ${corpo}`.trim())
  }
  const body = (await response.json()) as {
    banco_id: string | number
    atlas_public_key: string
    ambiente: string
    keyid?: string
  }
  if (!body.atlas_public_key) throw new Error("pareamento: resposta sem atlas_public_key")
  return {
    bancoId: String(body.banco_id),
    keyid: body.keyid ?? input.keyid,
    atlasPublicKey: body.atlas_public_key,
    ambiente: body.ambiente,
  }
}

export function integracaoPath(dataDir: string): string {
  return path.join(dataDir, "integracao.json")
}

export function loadIntegracao(dataDir: string, keys: BancoKeys): Identity | null {
  try {
    const raw = JSON.parse(readFileSync(integracaoPath(dataDir), "utf8")) as Integracao
    if (!raw.atlasPublicKey) return null
    return {
      atlasPublicKey: raw.atlasPublicKey,
      bancoPrivateKey: keys.privateKey,
      bancoPublicKey: keys.publicKey,
      bancoId: raw.bancoId,
      keyid: raw.keyid || keys.keyid,
    }
  } catch {
    return null
  }
}

export function saveIntegracao(dataDir: string, integracao: Integracao): void {
  mkdirSync(dataDir, { recursive: true })
  writeFileSync(integracaoPath(dataDir), JSON.stringify(integracao, null, 2) + "\n", "utf8")
}

export function identityFrom(keys: BancoKeys, integracao: Integracao): Identity {
  return {
    atlasPublicKey: integracao.atlasPublicKey,
    bancoPrivateKey: keys.privateKey,
    bancoPublicKey: keys.publicKey,
    bancoId: integracao.bancoId,
    keyid: integracao.keyid || keys.keyid,
  }
}

File: dotnet/Atlas.Averbacao/Security/Pairing.cs
using System.Net.Http.Json;
using System.Text.Json.Serialization;

namespace Atlas.Averbacao.Security;

public sealed record Integracao(string BancoId, string Keyid, string AtlasPublicKey, string Ambiente);

public static class Pairing
{
    public static async Task<Integracao> PairAsync(
        string atlasUrl,
        string token,
        string publicBaseUrl,
        string publicKey,
        string keyid,
        HttpClient? http = null)
    {
        var client = http ?? new HttpClient();
        var url = $"{atlasUrl.TrimEnd('/')}/v1/integracao/parear";
        using var req = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = JsonContent.Create(new PairBody(publicKey, publicBaseUrl, keyid)),
        };
        req.Headers.TryAddWithoutValidation("Authorization", $"Bearer {token}");
        var res = await client.SendAsync(req);
        var corpo = await res.Content.ReadAsStringAsync();
        if (!res.IsSuccessStatusCode)
            throw new InvalidOperationException($"pareamento falhou: HTTP {(int)res.StatusCode} {corpo}".Trim());
        var body = System.Text.Json.JsonSerializer.Deserialize<PairResponse>(corpo, Json.Options)
            ?? throw new InvalidOperationException("pareamento: resposta inválida");
        if (string.IsNullOrWhiteSpace(body.AtlasPublicKey))
            throw new InvalidOperationException("pareamento: resposta sem atlas_public_key");
        return new Integracao(
            body.BancoId ?? "",
            string.IsNullOrWhiteSpace(body.Keyid) ? keyid : body.Keyid!,
            body.AtlasPublicKey,
            body.Ambiente ?? "");
    }

    public static string IntegracaoPath(string dataDir) => Path.Combine(dataDir, "integracao.json");

    public static Identity? Load(string dataDir, BancoKeys keys)
    {
        try
        {
            var raw = File.ReadAllText(IntegracaoPath(dataDir));
            var integ = System.Text.Json.JsonSerializer.Deserialize<Integracao>(raw, Json.Options);
            if (integ is null || string.IsNullOrWhiteSpace(integ.AtlasPublicKey)) return null;
            return From(keys, integ);
        }
        catch
        {
            return null;
        }
    }

    public static void Save(string dataDir, Integracao integracao)
    {
        Directory.CreateDirectory(dataDir);
        File.WriteAllText(IntegracaoPath(dataDir), System.Text.Json.JsonSerializer.Serialize(integracao, Json.Options) + "\n");
    }

    public static Identity From(BancoKeys keys, Integracao integracao) => new(
        integracao.AtlasPublicKey,
        keys.PrivateKey,
        keys.PublicKey,
        integracao.BancoId,
        string.IsNullOrWhiteSpace(integracao.Keyid) ? keys.Keyid : integracao.Keyid);

    private sealed record PairBody(
        [property: JsonPropertyName("public_key")] string PublicKey,
        [property: JsonPropertyName("base_url")] string BaseUrl,
        [property: JsonPropertyName("keyid")] string Keyid);

    private sealed class PairResponse
    {
        [JsonPropertyName("banco_id")]
        public string? BancoId { get; set; }
        [JsonPropertyName("atlas_public_key")]
        public string? AtlasPublicKey { get; set; }
        [JsonPropertyName("ambiente")]
        public string? Ambiente { get; set; }
        [JsonPropertyName("keyid")]
        public string? Keyid { get; set; }
    }
}
